chain/network/src/peer_manager.rs

```diff
@@ -179,55 +236,135 @@ impl PeerManagerActor {
             monitor_peers_attempts: 0,
             started_connect_attempts: false,
             pending_update_nonce_request: HashMap::new(),
             network_metrics: NetworkMetrics::new(),
             edge_verifier_pool,
-            routing_table_pool: ibf_routing_pool,
+            routing_table_addr,
             txns_since_last_block,
             pending_incoming_connections_counter: Arc::new(AtomicUsize::new(0)),
             peer_counter: Arc::new(AtomicUsize::new(0)),
-            scheduled_routing_table_update: false,
             edge_verifier_requests_in_progress: 0,
-            #[cfg(feature = "test_features")]
-            adv_disable_edge_propagation: false,
-            #[cfg(feature = "test_features")]
-            adv_disable_edge_signature_verification: false,
-            #[cfg(feature = "test_features")]
-            adv_disable_edge_pruning: false,
+            adv_helper: AdvHelper::default(),
         })
     }
 
-    fn update_and_remove_edges(
+    fn update_routing_table_and_prune_edges(
         &mut self,
-        ctx: &mut Context<PeerManagerActor>,
-        can_save_edges: bool,
-        force_pruning: bool,
-        timeout: Duration,
+        ctx: &mut Context<Self>,
+        prune: Prune,
+        prune_edges_after: Duration,
     ) {
-        let edges_to_remove = self.routing_table.update(can_save_edges, force_pruning, timeout);
-        self.routing_table_pool
-            .send(RoutingTableMessages::RemoveEdges(edges_to_remove))
+        self.routing_table_addr
+            .send(RoutingTableMessages::RoutingTableUpdate { prune, prune_edges_after })
             .into_actor(self)
-            .map(|_, _, _| ())
+            .map(|response, act, _| match response {
+                Ok(RoutingTableMessagesResponse::RoutingTableUpdateResponse {
+                    edges_to_remove,
+                    peer_forwarding,
+                }) => {
+                    act.routing_table_view.remove_edges(&edges_to_remove);
+                    act.routing_table_view.peer_forwarding = peer_forwarding;
+                }
+                _ => error!(target: "network", "expected RoutingTableUpdateResponse"),
+            })
+            .spawn(ctx);
+    }
+
+    fn add_verified_edges_to_routing_table(
+        &mut self,
+        ctx: &mut Context<Self>,
+        edges: Vec<Edge>,
+        mode: AddEdgesMode,
+    ) {
+        if edges.len() == 0 {
+            return;
+        }
+        // RoutingTable keeps list of local edges; let's apply changes immediately.
+        for edge in edges.iter() {
+            if !(edge.contains_peer(&self.my_peer_id)) {
+                continue;
+            }
+            let key = edge.get_pair();
+            if !self.routing_table_view.is_edge_newer(&key, edge.nonce) {
+                continue;
+            }
+            self.routing_table_view
+                .local_edges_info
+                .insert((edge.peer0.clone(), edge.peer1.clone()), edge.clone());
```

Comment:
thought: curiously, this essentially duplicates `(peer0, peer1)` twice, as they are stored both as keys and as values. With some rust trickery, it should be possible to store just a `HashSet<Edge>`, which is then query-able using a `&(Peer, Peer)` pair (you'll need to implment `Borrow<(PeerId, PeerId)> for Edge`). Probably doesn't makes sense here, but a good trick to keep in mind.

Reply:
It's good to know that's it's possible. We could save even more memory by storing `Arc<PeerId>` and have an extra data structure, which removed duplicates.

For example `HashSet<Arc<PeerId>>` inside `RoutingTableActor`. I'm not sure if it's worth the effort.